File: 121. Best Time to Buy and Sell Stock.java
public class Solution {
    public int maxProfit(int[] prices) {
        int output = 0;
        int max = 0;
        
        for (int j = prices.length - 1; j >= 0; j--) {
            output = Math.max(output, max - prices[j]);
            max = Math.max(max, prices[j]);
        }
        
        return output;
    }
}

File: 121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        
        profit = 0
        minval = prices[0]
        
        for i in range(1, len(prices)):
            minval = min(minval, prices[i])
            profit = max(profit, prices[i] - minval)
        
        return profit
